trie/utils.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import "sync"
+
+// tracer tracks the changes of trie nodes. During the trie operations,
+// some nodes can be deleted from the trie, while these deleted nodes
+// won't be captured by trie.Hasher or trie.Committer. Thus, these deleted
+// nodes won't be removed from the disk at all. Tracer is an auxiliary tool
+// used to track all insert and delete operations of trie and capture all
+// deleted nodes eventually.
+//
+// The changed nodes can be mainly divided into two categories: the leaf
+// node and intermediate node. The former is inserted/deleted by callers
+// while the latter is inserted/deleted in order to follow the rule of trie.
+// This tool can track all of them no matter the node is embedded in its
+// parent or not, but valueNode is never tracked.
+type tracer struct {
+	lock   sync.RWMutex
+	insert map[string]struct{}
+	delete map[string]struct{}
+}
+
+// newTracer initializes trie node diff tracer.
+func newTracer() *tracer {
+	return &tracer{
+		insert: make(map[string]struct{}),
+		delete: make(map[string]struct{}),
+	}
+}
+
+// onInsert tracks the newly inserted trie node. If it's already
+// in the deletion set(resurrected node), then just wipe it from
+// the deletion set as it's untouched.
+func (t *tracer) onInsert(key []byte) {
+	// Tracer isn't used right now, remove this check later.
+	if t == nil {
+		return
+	}
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if _, present := t.delete[string(key)]; present {
+		delete(t.delete, string(key))
+		return
+	}
+	t.insert[string(key)] = struct{}{}
+}
+
+// onDelete tracks the newly deleted trie node. If it's already
+// in the addition set, then just wipe it from the addition set
+// as it's untouched.
+func (t *tracer) onDelete(key []byte) {
+	// Tracer isn't used right now, remove this check later.
+	if t == nil {
+		return
+	}
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if _, present := t.insert[string(key)]; present {
+		delete(t.insert, string(key))
+		return
+	}
```

Comment:
General question, really:
Is it faster to `check-presence-then-delete` than just `delete`?

Reply:
Answering my own question: 
```
BenchmarkDelete/10K-misses-del
BenchmarkDelete/10K-misses-del-6         	 3918998	       405.0 ns/op
BenchmarkDelete/10K-misses-check-del
BenchmarkDelete/10K-misses-check-del-6   	 6184447	       255.9 ns/op
```
On a map with 10K items, trying to delete items (which are not present) -- the `check-then-del` is indeed faster.